Additional Viewer Enhancements

dorodnic opened the pull request:
Adding simple distance measurement ability to the 3D view of the Viewer:
![Screenshot from 2020-05-19 06-19-11](https://user-images.githubusercontent.com/6958867/82314440-d2903900-9997-11ea-823c-5230feb680ca.png)

ev-mp commented on common/opengl3.cpp:
```diff
@@ -38,6 +38,7 @@ void vbo::upload(int attribute, const float* xyz, int size, int count, bool dyna
         dynamic ? GL_DYNAMIC_DRAW : GL_STATIC_DRAW);
     glVertexAttribPointer(attribute, size, GL_FLOAT, GL_FALSE, 0, 0);
     _size = count;
+    check_gl_error();
```
There is no legal note in the file's header

ev-mp commented on common/opengl3.cpp:
```diff
@@ -628,3 +682,44 @@ std::string fbo::get_status()
+            case GL_INVALID_ENUM:           error="INVALID_ENUM";           break;
+            case GL_INVALID_VALUE:          error="INVALID_VALUE";          break;
+            case GL_OUT_OF_MEMORY:          error="OUT_OF_MEMORY";          break;
+            case GL_INVALID_FRAMEBUFFER_OPERATION:  error="INVALID_FRAMEBUFFER_OPERATION";  break;
```
Default is missing

ev-mp commented on common/opengl3.cpp:
```diff
@@ -628,3 +682,44 @@ std::string fbo::get_status()
+
+        ss << "GL_" << error.c_str() << " - " << file << ":" << line << "\n";
+        err = glGetError();
+        has_errors = true;
```
`has_errors` will be inconsistent if in the previous line `err = GL_NO_ERROR`.
Also - all the retrieved values are overwritten by the last one . Is it by design?

ev-mp commented on common/opengl3.cpp:
```diff
@@ -628,3 +682,44 @@ std::string fbo::get_status()
+    {
+        err = glGetError();
+    }
+}
```
Missing EOF

ev-mp commented on common/opengl3.h:
```diff
@@ -7,6 +7,13 @@
+
+#define check_gl_error() _check_gl_error(__FILE__,__LINE__)
+
+#define MOUSE_PICK_USE_PBO     1                   // use pbo to improve performance, for debug only
```
Should it be leaved in place?

ev-mp commented on common/res/nx.h:
```diff
@@ -0,0 +1,6 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2020 Intel Corporation. All Rights Reserved.
+
+// This file is auto-generated from cubemap_nx.png
```
Do we need to specify the files (*.png) license/source?

ev-mp commented on common/skybox.cpp:
```diff
@@ -0,0 +1,114 @@
+    glBindTexture(GL_TEXTURE_2D, 0);
+
+    glDisable(GL_TEXTURE_2D);
+}
```
EOF

ev-mp commented on common/skybox.cpp:
```diff
@@ -0,0 +1,114 @@
+#include "skybox.h"
```
License label is missing

ev-mp commented on common/viewer.cpp:
```diff
@@ -388,249 +460,440 @@ namespace rs2
 
-        ImGui::PushStyleColor(ImGuiCol_Text, light_grey);
-        ImGui::PushStyleColor(ImGuiCol_TextSelectedBg, white);
+        // active = show_help_screen;
```
Can the commented code be removed?

ev-mp commented on common/viewer.cpp:
```diff
@@ -1783,13 +2118,24 @@ namespace rs2
+
+        glMatrixMode(GL_PROJECTION);
+        glPushMatrix();
+        gluPerspective(45, viewer_rect.w / win.framebuf_height(), 0.001f, 100.0f);
```
In case of T265 100 meter won't be enough  imho as you may want to see the whole path

ev-mp commented on common/viewer.cpp:
```diff
@@ -2033,28 +2473,294 @@ namespace rs2
+        {
+            if (distance < 0.01f)
+            {
+                label = to_string() << std::fixed << std::setprecision(3) << distance * 1000.f << " mm";
```
Nice

ev-mp commented on src/proc/zero-order.cpp:
```diff
@@ -516,6 +516,8 @@ namespace librealsense
-        return source.allocate_composite_frame(results);
+        if (results.size())
+            return source.allocate_composite_frame(results);
+        else return rs2::frame{};
```
Why not to pass through the input instead of an empty frame?

dorodnic commented:
Addressed code review comments and added some extra functionality - 
1. Finally found and fixed heap corruption on viewer close
2. SHIFT allows to continue measuring more than one segment
3. Measuring a closed loop will calculate and display area
4. Ctrl+Z can be used to undo changes
5. A bit improved mouse control, to avoid confusion during pan and zoom and highlight selected points & line segments
